### PR TITLE
[IMP] Auto transform valid url into link

### PR DESCRIPTION
**Pull request by sebgeelen**
Convert all url from pasted text.
Check and convert url in current Node when inserting a space.

see odoo task : https://www.odoo.com/web#action=333&active_id=1695&cids=1&id=2055864&menu_id=4720&model=project.task&view_type=form

### CODE REVIEW

File: src/editor.js [hunk 1]
```diff
@@ -940,9 +941,10 @@ export class OdooEditor extends EventTarget {
      */
     _applyRawCommand(method, ...args) {
         const sel = this.document.getSelection();
-        if (!this.editable.contains(sel.anchorNode) || (
-                sel.anchorNode !== sel.focusNode && !this.editable.contains(sel.focusNode))
-            ) {
+        if (
+            !this.editable.contains(sel.anchorNode) ||
+            (sel.anchorNode !== sel.focusNode && !this.editable.contains(sel.focusNode))
+        ) {
```
Comment (Goaman) [inline src/editor.js:947]: Is it a styling fix ? Could it go in another commit ?

File: src/utils/utils.js [hunk 2]
```diff
@@ -637,8 +637,10 @@ export function getDeepRange(editable, { range, sel, splitText, select, correctT
     // node, like the one resulting from a triple click, are corrected so
     // that it ends at the last position of the previous node instead.
     const beforeEnd = end.previousSibling;
-    if (correctTripleClick && !endOffset && (
-        !beforeEnd || (beforeEnd.nodeType === Node.TEXT_NODE && !isVisibleStr(beforeEnd)))
+    if (
+        correctTripleClick &&
+        !endOffset &&
+        (!beforeEnd || (beforeEnd.nodeType === Node.TEXT_NODE && !isVisibleStr(beforeEnd)))
```
Comment (Goaman) [inline src/utils/utils.js:643]: Is it a styling fix ? Could it go in another commit ?

File: src/editor.js [hunk 3]
```diff
@@ -1465,6 +1467,14 @@ export class OdooEditor extends EventTarget {
                     const range = selection.getRangeAt(0);
                     setCursor(range.endContainer, range.endOffset);
                 }
+                // check for url after user insert a space
```
Comment (Goaman) [inline src/editor.js:1470]: Check for url after user insert a space so that we don't transform an incomplete url.

File: src/editor.js [hunk 4]
```diff
@@ -1465,6 +1467,14 @@ export class OdooEditor extends EventTarget {
...
+                if (ev.data === ' ') {
+                    ev.preventDefault();
+                    this._checkForUrlInCurrentSelectionContainer();
+                    // const sel = this.document.getSelection();
+                    // setCursorEnd(sel.anchorNode);
```
Comment (Goaman) [inline src/editor.js:1475]: :eyes:

File: src/editor.js [hunk 5]
```diff
@@ -1716,13 +1726,78 @@ export class OdooEditor extends EventTarget {
...
+        }
+    }
+
+    /**
+     * Convert valid url text into links inside the given element
+     *
+     * @param {HTMLElement} el
+     */
+    _checkForUrlInElement(el) {
+        // We will not replace url inside already existing Link element
```
Comment (Goaman) [inline src/editor.js:1745]: // We will not replace url inside already existing Link element.

File: src/editor.js [hunk 6]
```diff
@@ -1716,13 +1726,78 @@ export class OdooEditor extends EventTarget {
...
+     */
+    _checkForUrlInCurrentSelectionContainer() {
+        const sel = this.document.getSelection();
+        if (sel && sel.anchorNode) {
+            this._checkForUrlInElement(closestElement(sel.anchorNode));
+        }
+    }
+
+    /**
+     * Convert valid url text into links inside the given element
```
Comment (Goaman) [inline src/editor.js:1740]: * Convert valid url text into links inside the given element.

File: src/editor.js [hunk 7]
```diff
@@ -1716,13 +1726,78 @@ export class OdooEditor extends EventTarget {
...
+        }
+    }
+
+    /**
+     * Create a Link in the node text based on the given data
+     *
+     * @param {Object} data
+     * @param {Node} textNode
+     */
+    _createLinkWithUrlInTextNode(data, textNode) {
```
Comment (Goaman) [inline src/editor.js:1770]: I would call the instruction directly rather than creating a function.

File: src/utils/utils.js [hunk 8]
```diff
@@ -898,6 +900,28 @@ export function getInSelection(document, selector) {
...
+ *
+ * @param {String} string
+ * @returns {Array}
+ */
+export function matchUrlIn(string) {
+    const re = /((https?:\/\/)?([^\s/?.#-]+\.)+[^\s/?.#-]+(\/[^\s]*)?)/gi;
+    let matches = [],
+        match;
+    while ((match = re.exec(string))) {
+        console.log('match', match);
```
Comment (Goaman) [inline src/utils/utils.js:914]: :eyes:

File: src/utils/utils.js [hunk 9]
```diff
@@ -898,6 +900,28 @@ export function getInSelection(document, selector) {
     );
 }
 
+/**
+ * Returns an array of urlData for url matched in the given string.
+ *
+ * @param {String} string
+ * @returns {Array}
+ */
+export function matchUrlIn(string) {
```
Comment (Goaman) [inline src/utils/utils.js:909]: getUrlsInfos

File: src/utils/utils.js [hunk 10]
```diff
@@ -24,6 +24,9 @@ export const CTGROUPS = {
     BR: CTYPES.BR,
 };
 
+export const UrlMatch = /((?:https?:\/\/)?(?:[a-z0-9-]{1,63}\.){1,2}[a-z]{2,15}(?:\/[^\s]*)?)/gi;
+export const UrlMatchWithInfos = /((https?:\/\/)?([a-z0-9-]{1,63}\.){1,2}[a-z]{2,15}(\/[^\s]*)?)/gi;
```
Comment (Goaman) [inline src/utils/utils.js:28]: 1. I would call it "url regex" rather than "url match".
2. For constants, we use all caps style: `URL_REGEX` and `URL_REGEX_WITH_INFOS`.

File: src/utils/utils.js [hunk 11]
```diff
@@ -898,6 +903,26 @@ export function getInSelection(document, selector) {
     );
 }
 
+/**
+ * Returns an array of url Infos for url matched in the given string.
```
Comment (Goaman) [inline src/utils/utils.js:907]: Infos without the cap `Infos` -> `infos`

File: src/utils/utils.js [hunk 12]
```diff
@@ -898,6 +903,26 @@ export function getInSelection(document, selector) {
...
+ *
+ * @param {String} string
+ * @returns {Array}
+ */
+export function getUrlsInfosIn(string) {
```
Comment (Goaman) [inline src/utils/utils.js:912]: `getUrlsInfosIn` ?

File: src/editor.js [hunk 13]
```diff
@@ -1716,13 +1725,64 @@ export class OdooEditor extends EventTarget {
         }
     }
 
+    /**
+     * Convert valid url text into links inside the given element.
+     *
+     * @param {HTMLElement} el
+     */
+    _checkForUrlInElement(el) {
```
Comment (Goaman) [inline src/editor.js:1733]: It's not a check, it is as converter as the docstring suggest. `_convertUrlInElement` ?

File: src/utils/utils.js (hunk 12 above)
Comment (sebgeelen) [reply src/utils/utils.js:912]: in the given string.

Comment (Goaman) [reply src/utils/utils.js]: Oups, I wanted to suggest: `getUrlsInfos`. I never saw a preposition at the end of the function name in Odoo.

File: src/utils/utils.js [hunk 14]
```diff
@@ -24,6 +24,9 @@ export const CTGROUPS = {
     BR: CTYPES.BR,
 };
 
+export const URL_REGEX = /((?:https?:\/\/)?(?:[a-z0-9-]{1,63}\.){1,2}[a-z]{2,15}(?:\/[^\s]*)?)/gi;
+export const URL_REGEX_WITH_INFOS = /((https?:\/\/)?([a-z0-9-]{1,63}\.){1,2}[a-z]{2,15}(\/[^\s]*)?)/gi;
```
Comment (EmilienD) [inline src/utils/utils.js:28]: I assume it's a deliberate choice to only check http/https protocoles ?

Comment (sebgeelen) [reply src/utils/utils.js:28]: Yes.
I don't think we want to auto transform into link other protocols. do we ?

File: src/editor.js [hunk 15]
```diff
@@ -1717,13 +1725,64 @@ export class OdooEditor extends EventTarget {
...
      */
     _onPaste(ev) {
         ev.preventDefault();
         const pastedText = (ev.originalEvent || ev).clipboardData.getData('text/plain');
-        this.execCommand('insertText', pastedText);
+        const splitAroundUrl = pastedText.split(URL_REGEX);
+        for (let i = 0; i < splitAroundUrl.length; i++) {
+            // Even indexes will always be plain text, and odd indexes will always be URL.
+            if (i % 2) {
+                const url = /^https?:\/\//gi.test(splitAroundUrl[i])
```
Comment (Zinston) [inline src/editor.js:1775]: You made a constant for `URL_REGEX`, why not make one for this one too? Again splitting hair but that way we prevent compiling it at every iteration of the loop.

File: src/editor.js [hunk 16]
```diff
@@ -1717,13 +1725,64 @@ export class OdooEditor extends EventTarget {
...
+                    ? splitAroundUrl[i]
+                    : 'https://' + splitAroundUrl[i];
+                this.execCommand(
+                    'insertHTML',
+                    '<a href="' + url + '" >' + splitAroundUrl[i] + '</a>',
```
Comment (Zinston) [inline src/editor.js:1780]: I believe we were instructed to use `target="_blank"` by default in Odoo, maybe it should be the case here too.

File: src/utils/utils.js (hunk 14 above)
Comment (Zinston) [inline src/utils/utils.js:28]: I'm just going to trust you on these regexes :eyes:

**Comment by Goaman** [comment]
It's safe to only deal with http/https because it covers most cases. If you
have a specific case, create the user can still create a proper link
through a modal.

File: src/utils/utils.js (hunk 14 above)
Comment (Goaman) [reply src/utils/utils.js:28]: It's safe to only deal with http/https because it covers most cases. If you have a specific case, create the user can still create a proper link through a modal.

File: src/utils/utils.js [hunk 17]
```diff
@@ -900,6 +903,26 @@ export function getInSelection(document, selector) {
...
+export function getUrlsInfosInString(string) {
+    let infos = [],
+        match;
+    while ((match = URL_REGEX_WITH_INFOS.exec(string))) {
+        infos.push({
+            url: match[2] ? match[0] : 'https://' + match[0],
+            label: match[0],
+            index: match.index,
+            length: match[0].length,
+        });
```
Comment (Zinston) [inline src/utils/utils.js:921]: Right so this would be the infamous `data` argument. In that case maybe a simple reference to this function in the docstring would be enough, though `data` and this differ in that `data` doesn't require `label` as the method is already getting a text node.

**Review by Zinston — COMMENTED** [review]
- 1 comment to cover a potential bug
- 1 comment to cover a potential feature requirement
The rest is me splitting hairs. Excellent job!

File: src/utils/utils.js (hunk 14 above)
Comment (Zinston) [reply src/utils/utils.js:28]: What I often see in editors is converting as soon as you have `\w+\.\w+` but that's how you end up with links whenever someone failed to put a space after a dot at the end of a sentence. This here is the other extreme where you _have_ to start you url with `http` for it to become a link, while in many cases you'd like to be able to simply type `www.site.com` or even `site.com`. Maybe the in-between solution is to cover `http`, `https` and `www.` ?

File: src/editor.js (hunk 15 above)
Comment (EmilienD) [reply src/editor.js:1775]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions

File: src/utils/utils.js (hunk 14 above)
Comment (Zinston) [reply src/utils/utils.js:28]: Sorry, I was replying to the comments here... then I looked at the regex and saw the `?`s. So here, the following bit will be converted right?

> I'm typing some text.And forgot to put a space.

-> `<a href="https://text.And">text.And</a>`

File: src/editor.js [hunk 18]
```diff
@@ -1466,6 +1468,12 @@ export class OdooEditor extends EventTarget {
                     const range = selection.getRangeAt(0);
                     setCursor(range.endContainer, range.endOffset);
                 }
+                // Check for url after user insert a space so we won't transform an incomplete url.
+                if (ev.data === ' ' && selection && selection.anchorNode) {
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply src/editor.js:1472]: hum, Good point. I was not aware we could have multiple char inside in the `ev.data`, but I guess mobile keyboard can do that ?

Il will change it ;)

File: src/editor.js [hunk 19]
```diff
@@ -1717,13 +1725,64 @@ export class OdooEditor extends EventTarget {
...
+        }
+
+        for (let child of el.childNodes) {
+            if (child.nodeType === Node.TEXT_NODE && child.length > 3) {
+                const childStr = child.nodeValue;
+                const matches = getUrlsInfosInString(childStr);
+                if (matches.length) {
+                    // We only to take care of the first match.
+                    // The method `_createLinkWithUrlInTextNode` will split the text node,
+                    // the other url matches will then be matched again in the nexts loops of el.childnodes.
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply src/editor.js:1746]: Well, mine was fine with It. it's lower than 80 char. should be fine I think...

File: src/editor.js (hunk 15 above)
Comment (Zinston) [reply src/editor.js:1775]: I was unaware, thanks.

File: src/editor.js [hunk 20]
```diff
@@ -1717,13 +1725,64 @@ export class OdooEditor extends EventTarget {
...
+                    this._createLinkWithUrlInTextNode(matches[0], child);
+                }
+            }
+        }
+    }
+
+    /**
+     * Create a Link in the node text based on the given data
+     *
+     * @param {Object} data
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply src/editor.js:1756]: good idea, I'll add more parameters.
It will be more clear indeed.

File: src/editor.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (Zinston) [reply src/editor.js:1746]: Fine by me :shrug:

File: src/editor.js (hunk 15 above)
Comment (sebgeelen) [reply src/editor.js:1775]: :+1:

File: src/editor.js (hunk 16 above)
Comment (sebgeelen) [reply src/editor.js:1780]: Didn't know that, I'll change it ;)

File: src/utils/utils.js (hunk 14 above)
Comment (sebgeelen) [reply src/utils/utils.js:28]: ;)

File: src/editor.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (Zinston) [reply src/editor.js:1472]: Yes, composition events can do that, so mobile keyboards and eg chinese input.

File: src/utils/utils.js (hunk 14 above)
Comment (EmilienD) [reply src/utils/utils.js:28]: I think mailto is pretty common but indeed it's an entirely reasonable choice to only match http/https, I just wanted to make sure it wasn't an oversight.

File: src/editor.js (hunk 16 above)
Comment (sebgeelen) [reply src/editor.js:1780]: After Discord discussion this will not be done here, maybe in a future dev with clearer specs.

File: src/utils/utils.js (hunk 14 above)
Comment (sebgeelen) [reply src/utils/utils.js:28]: > Sorry, I was replying to the comments here... then I looked at the regex and saw the `?`s. So here, the following bit will be converted right?
> 
> > I'm typing some text.And forgot to put a space.
> 
> -> `<a href="https://text.And">text.And</a>`
> 
> (Not that it matters extremely but just to be clear that we're doing what we want to be doing ^^).

yes Indeed.
